chronos/futures.nim

```diff
@@ -0,0 +1,369 @@
+#
+#                     Chronos
+#
+#  (c) Copyright 2015 Dominik Picheta
+#  (c) Copyright 2018-2023 Status Research & Development GmbH
+#
+#                Licensed under either of
+#    Apache License, version 2.0, (LICENSE-APACHEv2)
+#                MIT license (LICENSE-MIT)
+
+import std/[strutils, sequtils]
+import stew/base10
+import "."/[config, srcloc]
+export srcloc
+
+when chronosStackTrace:
+  when defined(nimHasStacktracesModule):
+    import system/stacktraces
+  else:
+    const
+      reraisedFromBegin = -10
+      reraisedFromEnd = -100
+
+when chronosStackTrace:
+  type StackTrace = string
+
+type
+  LocationKind* {.pure.} = enum
+    Create
+    Complete
+
+  CallbackFunc* = proc (arg: pointer) {.gcsafe, raises: [Defect].}
+
+  # TODO v4: consider removing this type and just use plain `CallbackFunc`
+  AsyncCallback* = object
+    function*: CallbackFunc
+    udata*: pointer
+
+  FutureState* {.pure.} = enum
+    Pending, Completed, Cancelled, Failed
+
+  FutureBase* = ref object of RootObj ## Untyped future.
+    when defined(chronosv4):
+      location: array[LocationKind, ptr SrcLoc]
+      callbacks: seq[AsyncCallback]
+      cancelcb: CallbackFunc
+      child: FutureBase
+      state: FutureState
+      error: ref CatchableError ## Stored exception
+      mustCancel: bool
+    else:
+      location*: array[LocationKind, ptr SrcLoc]
+      callbacks*: seq[AsyncCallback]
+      cancelcb*: CallbackFunc
+      child*: FutureBase
+      state*: FutureState
+      error*: ref CatchableError ## Stored exception
+      mustCancel*: bool
+
+    when chronosFutureId:
+      id*: uint
+
+    when chronosStackTrace:
+      errorStackTrace*: StackTrace
+      stackTrace*: StackTrace ## For debugging purposes only.
+
+    when chronosFutureTracking:
+      next*: FutureBase
+      prev*: FutureBase
+
+  # ZAH: we have discussed some possible optimizations where
+  # the future can be stored within the caller's stack frame.
+  # How much refactoring is needed to make this a regular non-ref type?
+  # Obviously, it will still be allocated on the heap when necessary.
+  Future*[T] = ref object of FutureBase ## Typed future.
+    when chronosStrictException:
+      when (NimMajor, NimMinor) < (1, 4):
+        closure*: iterator(f: Future[T]): FutureBase {.raises: [Defect, CatchableError], gcsafe.}
+      else:
+        closure*: iterator(f: Future[T]): FutureBase {.raises: [CatchableError], gcsafe.}
+    else:
+      closure*: iterator(f: Future[T]): FutureBase {.raises: [Exception], gcsafe.}
+    value*: T ## Stored value
+
+  FutureDefect* = object of Defect
+    cause*: FutureBase
+
+  FutureError* = object of CatchableError
+
+  CancelledError* = object of FutureError
+
+when defined(chronosV4):
+  type
+    NoValueError* = object of FutureError
+      ## Exception raised when trying to access the value of a future when
```

Comment:
I dont see any reasons to distinct this errors, it should be just a single error `FutureNotYetCompleted`, and it doesn't matter is it get accessed by `error` or by `read` procedures.

Reply:
changed to `FuturePendingError` - `Completed` would mean that it has a value (and not an error/cancel)

Reply:
Whatever Future has error or value it become completed, because it changes its state to state where it can't be changed anymore. So error should indicate the moment of time when attempt to get value or error from Future should fail with single type of Exception.

Reply:
currently, chronos is not precise when it comes to the names `Completed` and `Finished` - sometimes, `completed` means it has a `value` and sometimes it means its not pending (ie has value _or_ error) - in this PR, `Completed` only means "has value" and `Finished` means `not pending` but we can change that so that `Completed` means "has value".

Reply:
right - I see what you mean - so we have two possible error conditions:
* caller accesses `value` or `error` when future is pending
* caller accesses `error` when future completed successfully (so only the `value` field is valid - error contains `nil`

Do we want separate exceptions for the above two cases? In both cases, currently accessing `error` directly returns `nil`.

Reply:
667beafd104c1a68af5b8a2009cc5d4da1292905 pushed to show the problem / various states

Reply:
This problem is not only works for error, because you can try to call `Future.read()` on `Future[void]` too, it also represents situation when Future is completed but missing value/error. So i think we could introduce 2 exceptions
1. Signals when Future is not completed yet.
2. Signals that Future missing requested data, e.g. when `read()` called on `Future[void]` or `error()` called when there no error.

Reply:
latest version does exactly this for `read` and `readError` - `error()` is still unchanged because that's a larger breaking change - it did not raise an exception before, meaning this needs some careful thinking

Reply:
`readError()` which returns `nil` is impossible to use with `raise`, so even if this behavior is missing right now it will cause runtime problems, same idea with `read` on `void`.

Reply:
ok - I've made `error()` return `nil` if the future is not completed or completed with value for now - this is consistent with what happens today - we can return to this later before releasing v4.

`readError` instead raises an exception, so callers can choose if they want `nil` or an exception. I'm not too happy about the `readError` name.

Finally, in v4, `value()` now raises a defect if the future is in the wrong state - previously it returned the default value. Possibly, we should make `error()` raise a defect as well so as to "balance" the API in terms of defects.